### PR TITLE
pd_client: Do not reconnect on store tombstone error

### LINKED ISSUE #12506: PD client keeps reconnecting on error StoreTombstone
### What version of TiKV are you using?

master

### What operating system and CPU are you using?

No matter

### Steps to reproduce

Execute pd-ctl unsafe remove-failed-stores, and find 

### What did you expect?
No PD reconnection

### What did happened?\
PD client reconnects a lot of times
image
image

### PR DESCRIPTION
**Pull request by Connor1996**
### What is changed and how it works?

Issue Number: close #12506 

What's Changed:

```commit-message
do not reconnect PD on store tombstone error
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

### Release note 

```release-note
Fix the issue that keep retrying reconnecting PD when encounter store tombstone error
```

### CODE REVIEW

File: components/resolved_ts/src/advance.rs [hunk 1]
```diff
@@ -1,389 +0,0 @@
-// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
```
Comment (5kbpers) [inline components/resolved_ts/src/advance.rs:1]: Why remove this?

Comment (Connor1996) [reply components/resolved_ts/src/advance.rs]: PTAL again, it's removed by mistake

File: components/resolved_ts/src/advance.rs [hunk 2]
```diff
@@ -234,7 +234,10 @@ pub async fn region_resolved_ts_store(
                 let client =
                     get_tikv_client(to_store, pd_client, security_mgr, env, tikv_clients.clone())
                         .await
-                        .map_err(|e| (to_store, true, format!("[get tikv client] {}", e)))?;
+                        .map_err(|e| {
+                            let reconnect = !matches!(e, pd_client::Error::StoreTombstone(..));
```
Comment (BusyJay) [inline components/resolved_ts/src/advance.rs:238]: Why not reused `should_not_retry`?

File: components/pd_client/src/util.rs [hunk 3]
```diff
@@ -441,8 +441,13 @@ where
     fn should_not_retry(resp: &Result<Resp>) -> bool {
         match resp {
             Ok(_) => true,
-            // Error::Incompatible is returned by response header from PD, no need to retry
-            Err(Error::Incompatible) => true,
+            // these errors are not caused by network, no need to retry
+            Err(
+                Error::Incompatible
```
Comment (BusyJay) [inline components/pd_client/src/util.rs:446]: How about make the matching as a method of `Error`?

File: components/pd_client/src/errors.rs [hunk 4]
```diff
@@ -29,10 +27,22 @@ pub enum Error {
 
 pub type Result<T> = result::Result<T, Error>;
 
+impl Error {
+    pub fn retryable(&self) -> bool {
+        match self {
+            Error::Grpc(_) | Error::Other(_) | Error::ClusterNotBootstrapped(_) => true,
+            Error::RegionNotFound(_)
```
[the remark this answers is not in the document]
Comment (Connor1996) [reply components/pd_client/src/errors.rs:34]: call `get_region_and_leader` but the region heartbeat hasn't been received yet

File: components/pd_client/src/errors.rs [hunk 5]
```diff
@@ -29,10 +27,22 @@ pub enum Error {
...
+            | Error::StoreTombstone(_)
+            | Error::GlobalConfigNotFound(_)
```
Comment (BusyJay) [inline components/pd_client/src/errors.rs:36]: Why this doesn't need retry?

Comment (Connor1996) [reply components/pd_client/src/errors.rs:36]: the config should be wrong, it doesn't the result after retry
/cc @glorv

Comment (glorv) [reply components/pd_client/src/errors.rs:36]: `GlobalConfigNotFound` is returned when the target key does not exist. So I think it is reasonable to not retry this error as the retry is likely to return the same error.

Though, In the current implement, the tikv side code just ignores this error with an error log, so there is no chance to propgate this error the outer level error handling code.

https://github.com/tikv/tikv/blob/c461cd829fc6e06f071441f70645c8403e7453a5/components/pd_client/src/client.rs#L309-L318

**Review by BusyJay — APPROVED** [review]

**Pull request merged**